tests/code_update.py

```diff
@@ -51,6 +51,180 @@ def test_verify_quotes(network, args):
+
+    LOG.info(f"Checking governance table security policies")
+    expected = [
+        {
+            "digest": DEFAULT_SNP_SECURITY_POLICY_DIGEST,
+            "raw": DEFAULT_SNP_SECURITY_POLICY,
+        }
+    ]
+    expected.sort(key=lambda x: x["digest"])
+    assert policies == expected, [(a, b) for a, b in zip(policies, expected)]
+
+    LOG.info(f"Checking node joined with a blank security policy")
+    # TODO: Do this
+
+    network.consortium.retire_security_policy(primary, DEFAULT_SNP_SECURITY_POLICY_DIGEST)
+    network.consortium.add_new_security_policy(primary, DEFAULT_SNP_SECURITY_POLICY, DEFAULT_SNP_SECURITY_POLICY_DIGEST)
+
+@reqs.description("Node with no security policy set but good digest not matching ledger joins successfully")
+def test_add_node_with_no_raw_security_policy_not_matching_ledger(network, args):
+
+    if not IS_SNP:
+        LOG.warning("Skipping test_add_node_with_correct_security_policy with non SNP platform")
+        return network
+
+    replacement_package = (
+        "samples/apps/logging/liblogging"
+        if args.package == "libjs_generic"
+        else "libjs_generic"
+    )
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+    new_node = network.create_node("local://localhost")
+    network.join_node(new_node, replacement_package, args, timeout=3, env=dict())
+    network.trust_node(new_node, args)
+    primary, others = network.find_nodes()
+    nodes = [primary, *others]
+    assert len(nodes) == len(prev_nodes) + 1, "New node not added to network"
+
+    with primary.client() as client:
+        r = client.get("/gov/security_policy")
+        policies = sorted(r.body.json()["policies"], key=lambda x: x["digest"])
+
+    LOG.info(f"Checking governance table security policies")
+    expected = [
+        {
+            "digest": DEFAULT_SNP_SECURITY_POLICY_DIGEST,
+            "raw": DEFAULT_SNP_SECURITY_POLICY,
+        }
+    ]
+    expected.sort(key=lambda x: x["digest"])
+    assert policies == expected, [(a, b) for a, b in zip(policies, expected)]
+
+    LOG.info(f"Checking node joined with a blank security policy")
+    # TODO: Do this
+
+@reqs.description("Node where raw security policy doesn't match digest fails to join")
+def test_add_node_with_mismatched_security_policy_digest(network, args):
+
+    if not IS_SNP:
+        LOG.warning("Skipping test_add_node_with_mismatched_security_policy_digest with non SNP platform")
+        return network
+
+    replacement_package = (
+        "samples/apps/logging/liblogging"
+        if args.package == "libjs_generic"
+        else "libjs_generic"
+    )
+
+    primary, others = network.find_nodes()
+    prev_nodes = [primary, *others]
+    new_node = network.create_node("local://localhost")
+    try:
+        network.join_node(new_node, replacement_package, args, timeout=3, env={"SECURITY_POLICY": "eyJhbGxvd19hbGwiOmZhbHNlLCJjb250YWluZXJzIjp7Imxlbmd0aCI6MCwiZWxlbWVudHMiOm51bGx9fQ=="})
```

Comment:
```suggestion
        network.join_node(new_node, replacement_package, args, timeout=3, env={"SECURITY_POLICY": "invalid_security_policy"})
```

So that it's obvious to the reader what the test is doing.

Reply:
If we do that, joining fails during converting to a string rather than the check to ensure the raw policy matches the digest, so we wouldn't have that coverage of that check. Of course, if the base64 doesn't parse then it also shouldn't match so I can see the argument not to worry about that?

Reply:
Good point. Would this work if we did `b64encode(b"invalid_security_policy").decode()` (valid base64 string but clear to the reader what the test does)?

Reply:
Yeah that's a great idea